tests/test_util/display.rs

```diff
@@ -0,0 +1,166 @@
+use chalk_integration::{program::Program, query::LoweringDatabase, tls};
+use chalk_solve::display::{self, WriterState};
+use regex::Regex;
+use std::{fmt::Debug, sync::Arc};
+
+macro_rules! reparse_test {
+    (program $program:tt) => {
+        crate::test_util::display::reparse_test(crate::test_util::strip_leading_trailing_braces(
+            stringify!($program),
+        ))
+    };
+    (program $program:tt produces $diff:tt) => {
+        crate::test_util::display::reparse_into_different_test(
+            crate::test_util::strip_leading_trailing_braces(stringify!($program)),
+            crate::test_util::strip_leading_trailing_braces(stringify!($diff)),
+        )
+    };
+    (program $program:tt formatting matches $res:literal) => {
+        crate::test_util::display::test_formatting(
+            crate::test_util::strip_leading_trailing_braces(stringify!($program)),
+            $res,
+        )
+    };
+}
+
+pub fn write_program(program: &Program) -> String {
+    let mut out = String::new();
+    let ws = &WriterState::new(program);
+    for datum in program.adt_data.values() {
+        display::write_top_level(&mut out, ws, &**datum).unwrap();
+    }
+    for datum in program.trait_data.values() {
+        display::write_top_level(&mut out, ws, &**datum).unwrap();
+    }
+    for datum in program.impl_data.values() {
+        display::write_top_level(&mut out, ws, &**datum).unwrap();
+    }
+    for datum in program.opaque_ty_data.values() {
+        display::write_top_level(&mut out, ws, &**datum).unwrap();
+    }
+    out
+}
+
+fn program_diff(original: &impl Debug, produced: &impl Debug) -> String {
```

Comment:
Wait are we actually diffing the program text? I don't think this is really helpful. Moreso, we just want the goals to be solvable (or fail) in the same way.

Reply:
It's diffing the `Debug` output of `Program`! I'll add some comments to it at least. This particular function is just nice output if we fail - it isn't a test. The actual test _is_ a `old_lowered_program == new_lowered_program` test, though.

We decided to use an exact comparison between `Program`s because it allowed us to really easily write a bunch of different tests without having to specify goals for every one, and that let us cover a lot more ground with less test code. With this, we can just test that a struct with fields outputs the same struct with fields, without having to figure out what exactly in chalk uses fields and to rely on that code as well.

It seems reasonable that the end goal is to keep goals solveable the same way. To that end, I'd consider these as unit tests. They test the exact behavior of this specific `display` code, and then the `logging_db` tests can assume that we render items correctly, and can focus on ensuring we get everything necessary for goals to pass (or fail).

Does that seem reasonable? We could also revise this testing strategy, but it would mean rewriting most of the `display` tests so that they have reasonable goals to follow.

Reply:
Yeah, honestly the test coverage is really good. And it seems pretty reasonable to just check that `old_lowered_program == new_lowered_program`. I was originally skeptical about this, because I thought it would end up being dificult. But it doesn't seem so.